### PR TITLE
feat: providers and models configuration

### LINKED ISSUE #46: SIM-Config providers and models in settings screen

### LINKED ISSUE #415: SIM-Providers are shown even if they are not configured
# Context

When starting a fresh install or boot of the simulator, all default providers are shown even if they are not configured in the backend
image

# Impact

Bad UX: This can be very counter intuitive for a user who creates a validator and then finds out through errors that the LLM provider is not configured in the backend

# Extra
Highly related to https://github.com/yeagerai/genlayer-simulator/issues/359

# Solution:
- Check availability for providers and models within the call to get_providers..
- Show unavailability in UI and educate user as to why

### PR DESCRIPTION
**Pull request by denishacquin**
Fixes #46 
Fixes #415 

# What

- Added ability to create / update providers from the frontend
  - Added provider list in settings
  - Client-side validation of configs
  - Client-side validation of unique constraints
  - Handling of misconfiguration cases in backend and UI
- Reworked the navigation to improve structure
  - New dedicated Validators tab
  - Profile becomes Settings
- Fixed a bug with a broken condition
- Fixed: selecting an account now closes the dropdown
- Fixed: integer validation on stake field that was breaking e2e test
- Moved account creation/delete directly in the account dropdown for simplification
- Fixed a scrolling issue on the home tab that was sometimes causing e2e tests to fail

Modals:
image
image
Nav:
image
image

# Why

- to allow developers to manage their inference providers
- to improve navigation

# Testing done

- tested the new features
- tested the bug fixes
- e2e frontend OK
- unit frontend OK
- backend integration OK

# Decisions made

- Added unique constraint on provider+model+plugin in postrgres
- Added AJV to validate json schema client-side
- Added lodash for easy grouping/sorting in JS
- Skipping two fields on the frontend for now (they don't create errors however because they have a default value in the schema). For future reference, I added a warning in the console to be able to spot missing fields easily:
image

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

- Try adding / editing providers, then adding / editing validators

# User facing release notes

- Added the ability to add and manage inference providers from in the settings
- Reworked navigation
- Move account managing to the account select menu
- Fixed other various bugs and issues

### CODE REVIEW

File: backend/node/create_nodes/providers_schema.json [hunk 1]
```diff
@@ -191,7 +192,8 @@
                 "default": 0
               },
               "stop": {
-                "const": ""
+                "const": "",
+                "default": ""
               },
```
[the remark this answers is not in the document]
Comment (denishacquin) [reply backend/node/create_nodes/providers_schema.json:197]: But kept field hidden in ui however.

File: backend/database_handler/llm_providers.py [hunk 2]
```diff
@@ -25,10 +27,23 @@ def get_all(self) -> list[LLMProvider]:
         ]
 
     def get_all_dict(self) -> list[dict]:
-        return [
-            _to_domain(provider).__dict__
-            for provider in self.session.query(LLMProviderDBModel).all()
-        ]
+        ## TODO: double check this with agustin
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply backend/database_handler/llm_providers.py:30]: looks great!

**Review by denishacquin — COMMENTED** [review]
self reviewed

File: backend/protocol_rpc/endpoints.py [hunk 3]
```diff
@@ -127,7 +128,7 @@ def create_validator(
     # fallback for default provider
     # TODO: only accept all or none of the config fields
     llm_provider = None
-    if not (config and plugin and plugin_config):
+    if not (config is None and plugin is None and plugin_config is None):
```
Comment (cristiam86) [inline backend/protocol_rpc/endpoints.py:131]: @denishacquin this is not what the comment says. This condition is equal to:
if config != None or plugin != None or plugin_config != None

**Review by cristiam86 — COMMENTED** [review]
I left a small comment and I will approve once I checked&tested the branch.

Would it be possible to add some sort of tests to the new screen?

As a side note, it is kind of confusing that Provider is the name for a provider and a model. Next time let's try to solve this issues before they are not worth it like right now.

**Comment by denishacquin** [comment]
> As a side note, it is kind of confusing that Provider is the name for a provider and a model. Next time let's try to solve this issues before they are not worth it like right now.

I agree which is why I was playing around with "**Provider Config**". It is true that the entity we are talking about is a kind of preset so to speak, that combines provider/model/plugin and default config.

Totally open to suggestions, but like you said, it's not dealbreaker right now and since it's an advanced feature, we should be fine for now indeed.

**Comment by denishacquin** [comment]
added migration to remove legacy heuristai providers

**Comment by denishacquin** [comment]
image
image
image
image

@cristiam86 @AgustinRamiroDiaz made some copy adjustments, added tooltips, added a link to settings from validator modal.

I think this helps with user education, WDYT?

File: backend/protocol_rpc/endpoints.py (hunk 3 above)
Comment (denishacquin) [reply backend/protocol_rpc/endpoints.py]: fixed this, thanks for the spot!

File: frontend/test/e2e/ManageProviders.spec.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    // Create
+    await settingsPage.createCustomProvider(providerParams);
+
+    const newProviders = await settingsPage.getProvidersElements();
+
+    expect(
+      newProviders.length,
+      'number of providers should be greather than old providers list',
+    ).be.greaterThan(initialProviders.length);
```
Comment (cristiam86) [inline frontend/test/e2e/ManageProviders.spec.ts:44]: @denishacquin we should be more specific and expecting the exact number of providers after the creation

File: frontend/test/e2e/ManageProviders.spec.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      By.xpath("//button[@data-testid='provider-item-confirm-delete']"),
+    );
+    // call delete validator button
+    await confirmDeleteProviderBtn.click();
+
+    const providers = await driver.findElements(
+      By.xpath("//button[@data-testid = 'provider-item']"),
+    );
+
+    expect(
```
Comment (cristiam86) [inline frontend/test/e2e/ManageProviders.spec.ts:106]: @denishacquin same as before, please be more specific

**Review by cristiam86 — APPROVED** [review]
Only minor things in e2e tests

**Review by cristiam86 — APPROVED** [review]

**Pull request merged**